### PR TITLE
Cache

### PR DESCRIPTION
**Pull request by MioQuispe**
This caches entities by their mask

### CODE REVIEW

File: lib/entity_manager.js [hunk 1]
```diff
@@ -37,6 +40,8 @@ EntityManager.prototype.create = function() {
   var entity = this._entityInst[id]
   this._entityFlag[id] = ENTITY_ALIVE
 
+  entity._cacheIndex = this._entityCache[0].push(entity) - 1
```
Comment (ooflorent) [inline lib/entity_manager.js]: Do not add anything to the entity.
Instead create an `Int32Array` to store indexes like `_entityMask`.

File: lib/entity_manager.js [hunk 2]
```diff
@@ -104,7 +120,32 @@ EntityManager.prototype._get = function(id, C) {
...
+  var entity = this._entityInst[id]
+  var cache = this._entityCache[this._entityMask[id]]
+  if (cache.length === 1 || entity._cacheIndex === cache.length - 1) {
+    cache.pop()
+  } else {
+    var lastCacheElement = cache.pop()
+    lastCacheElement._cacheIndex = entity._cacheIndex
+    cache[lastCacheElement._cacheIndex] = lastCacheElement
+  }
+  entity._cacheIndex = false
```
Comment (ooflorent) [inline lib/entity_manager.js]: Do not mix types. Cache index must be an integer.

**Comment by MioQuispe** [comment]
Fyi, I tried preallocating an array based on the amount of entities the query will return. The speedup was roughly 2-3x for 100k entities 50k returned

File: lib/entity_manager.js (hunk 1 above)
Comment (MioQuispe) [inline lib/entity_manager.js]: Good idea.

File: lib/entity_manager.js (hunk 2 above)
Comment (MioQuispe) [inline lib/entity_manager.js]: True. it isn't necessary to make it false.

**Comment by ooflorent** [comment]
Can you rebase on master?
I've fixed a bug on `Entity#destroy`

**Comment by MioQuispe** [comment]
Sure thing.

File: lib/entity_manager.js [hunk 3]
```diff
@@ -22,6 +22,9 @@ function EntityManager(components, storage) {
 
   this._entityPool = []
   this._entityCounter = 0
+
+  this._entityCache = {}
+  this._entityCache[0] = []
```
Comment (ooflorent) [inline lib/entity_manager.js:29]: Can be optimized:

``` js
this._entityCache = FastMap()
this._entityCache[0] = []
```

See https://gist.github.com/ooflorent/c58c9a0f08923973a5e3

File: lib/entity_manager.js [hunk 4]
```diff
@@ -104,7 +120,32 @@ EntityManager.prototype._get = function(id, C) {
...
+
+EntityManager.prototype._cacheRemove = function(id) {
+  var entity = this._entityInst[id]
+  var cache = this._entityCache[this._entityMask[id]]
+  if (cache.length === 1 || entity._cacheIndex === cache.length - 1) {
+    cache.pop()
+  } else {
+    var lastCacheElement = cache.pop()
+    lastCacheElement._cacheIndex = entity._cacheIndex
+    cache[lastCacheElement._cacheIndex] = lastCacheElement
```
Comment (ooflorent) [inline lib/entity_manager.js]: What if `lastCacheElement` is `entity`?
Does it work as expected?

Comment (MioQuispe) [inline lib/entity_manager.js]: `entity._cacheIndex === cache.length - 1`
This is the check for it

Comment (MioQuispe) [inline lib/entity_manager.js]: I will add tests to check that it returns the correct entities though.

File: lib/entity_manager.js [hunk 5]
```diff
@@ -55,9 +60,12 @@ EntityManager.prototype.query = function() {
   }
 
   var result = []
-  for (var id = 0; id < this._entityCounter; id++) {
-    if (this._entityFlag[id] === ENTITY_ALIVE && (this._entityMask[id] & mask) === mask) {
-      result.push(this._entityInst[id])
+
+  var entityMasks = Object.keys(this._entityCache)
```
Comment (ooflorent) [inline lib/entity_manager.js:65]: This can be a real performance killer.
If we have 32 components it may generate a lot of different masks.

File: lib/entity_manager.js (hunk 3 above)
Comment (MioQuispe) [inline lib/entity_manager.js:29]: So do you suggest I add fast.js to the project and require it?

**Comment by ooflorent** [comment]
Well let's finish the PR, add some tests to ensure there a no duplicates and we should be good

File: lib/entity_manager.js (hunk 3 above)
Comment (ooflorent) [inline lib/entity_manager.js:29]: Yes. Next to `utils.js`.

File: lib/entity_manager.js (hunk 5 above)
Comment (MioQuispe) [inline lib/entity_manager.js:65]: Yes, I will add something that removes the empty ones.

File: lib/entity_manager.js (hunk 3 above)
Comment (MioQuispe) [inline lib/entity_manager.js:29]: Getting:

```
/home/user/Projects/mq_makr/makr/lib/fast.js:10
  ForceEfficientMap.prototype = self
  ^^^^^^^^^^^^^^^^^
SyntaxError: Unexpected identifier
```

Comment (ooflorent) [inline lib/entity_manager.js:29]: Fixed on the gist.

Comment (MioQuispe) [inline lib/entity_manager.js:29]: Oh, wow didn't notice the missing brackets

File: lib/entity_manager.js (hunk 5 above)
Comment (MioQuispe) [inline lib/entity_manager.js:65]: Want me to fix this before the merge?

Comment (ooflorent) [inline lib/entity_manager.js:65]: Sure! But I won't merge it immediately. I want to add more tests and benchmarks.

Comment (MioQuispe) [inline lib/entity_manager.js:65]: I tried two approaches so far, changing the property descriptor to enumerable: false so it doesn't appear in Object.keys(). This resulted in quite bad performance when adding/removing components.

Also tried simply deleting the property which I know is a bad thing to do on objects. But the performance is much better. 10x faster when adding components, 100x faster when removing them.

Overall it increased the query performance by roughly 1.5-2x on 1000 entities.

Any other ideas on how to do this?

**Comment by MioQuispe** [comment]
I've added several tests which seem to confirm it's working correctly. But I have something I want to still add to it which would effectively get around the .push() problem. I predict this will have a very big impact on the performance.

**Comment by MioQuispe** [comment]
Initial numbers (All tests pass, including the new ones):

```
                  EntityManager
  57,264,548 op/s » query on 1000 entities with 1 component (500 matches)
  35,554,763 op/s » query on 1000 entities with 3 components (84 matches)
  10,677,708 op/s » query on 1000 entities with 10 components (1 match)

                  EntityManager
  59,946,961 op/s » query on 10000 entities with 1 component (~5000 matches)
  32,865,792 op/s » query on 10000 entities with 3 components (~840 matches)
   8,758,721 op/s » query on 10000 entities with 10 components (~4 matches)

                  EntityManager
  58,063,592 op/s » query on 100000 entities with 1 component (~50000 matches)
  32,834,461 op/s » query on 100000 entities with 3 components (~8400 matches)
   8,797,725 op/s » query on 100000 entities with 10 components (~40 matches)
```

As you see, it will be independent of the amount of entities.
So what's going on? The query results are cached.
Whenever an entity is created/destroyed/has a component added/removed
the respective query result caches are set as invalid and get rebuilt the next time query is called with their mask. So no, these numbers aren't 100% realistic as each affected query cache will have to be rebuilt whenever an entity is modified. The good thing is (apart from the crazy speeds when the query cache is used) this should mean much less work for the garbage collector and we aren't really using any extra memory since these arrays would be created anyway each time a query is performed. Unfortunately this adds quite a bit of overhead for modifying entities:

Before:
     168,066,626 op/s » #destroy
      81,404,501 op/s » #add
      84,897,373 op/s » #remove
       1,533,740 op/s » #create

After:
      34,618,246 op/s » #destroy
       2,773,040 op/s » #add
       3,422,823 op/s » #remove
         880,812 op/s » #create

This is something I'd like to fix by scheduling the cache invalidations for the next query somehow. I'd like to see the numbers back to the old ones ;). Let me know what you think. All the other numbers stayed the same btw.

**Comment by eranimo** [comment]
Can this be merged?

**Comment by ooflorent** [comment]
@eranimo I don't think so. I'm not fully satisfied by this change. While the performance improvement is huge for `query`, there is a significant overhead for entity manipulation methods.
